### PR TITLE
Honor masking of systemd-sysctl.service

### PR DESCRIPTION
**Pull request by zeha**
During package install on systemd-based systems, some sysctl settings
should be set (e.g. vm.max_map_count).

In some environments, changing sysctl settings plainly does not work;
previously a global environment variable named ES_SKIP_SET_KERNEL_PARAMETERS
was introduced to skip calling sysctl, but this causes trouble for:
 * configuration management systems, which usually cannot apply an env
   var when running a package manager
 * package upgrades, which will not have the env var set any more, and
   thus leaving the package management system in a bad state (possibly
   half-way upgraded, can be very hard to recover)

This removes the env var again and instead of calling systemd-sysctl manually,
tells systemd to restart the wrapper unit - which itself can be masked by
system administrators or management tools if it is known that sysctl does
not work in a given environment.
The restart is not silent on systems in their default configuration, but
is ignored if the unit is masked.

Related: #21899, elastic/puppet-elasticsearch#806

### CODE REVIEW

**Comment by jasontedor** [comment]
In the course of doing my due diligence on this one, I've learned that we are violating packaging guidelines by allowing the installation scripts to exit with non-zero status code. In the [bluntest possible terms](https://fedoraproject.org/wiki/Packaging:Scriptlets):

> All scriptlets MUST exit with the zero exit status.

Technically this is a breaking change so I will integrate this and add a note to the migration docs.

Thank you for this PR @zeha!

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by jasontedor** [comment]
I pushed 2f8ba3ad39af7b47ee7b04051998469c359a31d7 to add a note to the migration docs.

**Comment by pdf** [comment]
This breaks deb systems where the `systemd-sysctl.service` is masked, because packaging scripts are implicitly run with `-e`, so the non-zero exit code from the call to restart a masked service aborts package installation.

File: distribution/src/main/packaging/scripts/postinst [hunk 1]
```diff
@@ -48,17 +48,8 @@ case "$1" in
...
-    elif command -v /lib/systemd/systemd-sysctl > /dev/null; then
-        /lib/systemd/systemd-sysctl
-    fi
-elif [ "$ES_SKIP_SET_KERNEL_PARAMETERS" == "true" ]; then
-    echo "skipping setting kernel parameters"
-else
-    echo "unrecognized value [$ES_SKIP_SET_KERNEL_PARAMETERS] for ES_SKIP_SET_KERNEL_PARAMETERS; must [false] (default) or [true]"
-    exit 1
+if command -v systemctl > /dev/null; then
+    systemctl restart systemd-sysctl.service
```
Comment (pdf) [inline distribution/src/main/packaging/scripts/postinst]: If this call is meant to fail silently, it should be:
```
systemctl restart systemd-sysctl.service || true
```

**Comment by porbas** [comment]
When trying to upgrade ES on ubuntu 16.04:
```
root@runner1-gitlab-prd:~# root@runner1-gitlab-prd:~# apt install -f
Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 7 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Setting up elasticsearch (5.5.0) ...
Failed to restart systemd-sysctl.service: Unit systemd-sysctl.service is masked.
dpkg: error processing package elasticsearch (--configure):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
 elasticsearch
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

change in postinst script, line 57 (add trailing "|| true"):

      systemctl restart systemd-sysctl.service || true

Now it works! :)
```
root@runner1-gitlab-prd:~# apt install -f
Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 7 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Setting up elasticsearch (5.5.0) ...
Failed to restart systemd-sysctl.service: Unit systemd-sysctl.service is masked.
```

**Comment by pdf** [comment]
ping @jasontedor your currently live packages are borken.

**Comment by jasontedor** [comment]
Sorry for this. I tested the PR on my workstation which runs Fedora 25 and it does not behave this way; it's only Debian based systems that see this problem. I opened #25657.

**Comment by juergenknaack** [comment]
Taking a look at #25657, i think adding /usr/lib/sysctl.d/elasticsearch.conf is breaking systemd-sysctl in the first place on systems, where this setting is not allowed/possible (e.g. container).
Masking the service as a consequence and working around it within elasticsearch installer may fix it for elasticsearch but still would leave it broken to others.
There has to be a way to avoid putting "vm.max_map_count=262144" into /usr/lib/sysctl.d/elasticsearch.conf is the proper way for these cases.

**Comment by pdf** [comment]
@juergenknaack not necessarily saying that this is a non-problem, but your unprivileged containers really should be masking `systemd-sysctl.service` by default, since it will never work, and other source may attempt to populate sysctl.  This sounds more like a container infrastructure problem than an elasticsearch packaging problem.

**Comment by juergenknaack** [comment]
@pdf I see where you are coming from... If there is no reasonable usecase for systemd-sysctl in containers at all, masking would be an appropriate way. Agree to that.

**Comment by juergenknaack** [comment]
@pdf I rechecked our container environment (Virtuozzo). We use systemd-sysctl to enforce some settings (e.g. net.ipv4.ip_forward=1) and it's working for us. The additonal /usr/lib/sysctl.d/elasticsearch file breaks this.

**Comment by pdf** [comment]
@juergenknaack looks like the way lxc handles this is to add start condition overrides to `systemd-sysctl.service` that check whether `/proc/sys/net` specifically is writable (rather than `/proc/sys`), then sets it to run `systemd-sysctl` with `--prefix net` added, to restrict it to only apply properties for `net`.  This is absolutely out of scope for ES to fix - you should contact the VZ or distro maints.

**Comment by jasontedor** [comment]
Thank you @pdf, you have it exactly right.

**Pull request merged**